small/mempool_asan.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "mempool.h"
+
+enum {
+	/** Maximum alignment for the ASAN mempool allocation. */
+	MEMPOOL_ASAN_MAX_ALIGNMENT = 4096,
+};
+
+void
+mempool_create(struct mempool *pool, struct slab_cache *cache,
+	       uint32_t objsize)
+{
+	(void)cache;
+	small_asan_assert(objsize > 0);
+	pool->objsize = objsize;
+	pool->objcount = 0;
+
+	/* Alignment is explained in header. */
+	pool->alignment = 1 << (__builtin_ffs(objsize) - 1);
+	if (pool->alignment > MEMPOOL_ASAN_MAX_ALIGNMENT)
+		pool->alignment = MEMPOOL_ASAN_MAX_ALIGNMENT;
+}
+
+void
+mempool_destroy(struct mempool *pool)
+{
+	/* Safety and also makes mempool_is_initialized work. */
+	memset(pool, 0, sizeof(*pool));
+}
+
+void *
+mempool_alloc(struct mempool *pool)
+{
+
+	struct mempool_object *obj =
+			small_asan_alloc(pool->objsize, pool->alignment, 0);
+	pool->objcount++;
+
+	return small_asan_payload_from_header(obj);
+}
+
+SMALL_NO_SANITIZE_ADDRESS void
+mempool_free(struct mempool *pool, void *ptr)
+{
+	struct mempool_object *obj = small_asan_header_from_payload(ptr);
+	pool->objcount--;
+
+	small_asan_free(obj);
```

Comment:
Let's add a check that ensures the the object was allocated from the pool.

Reply:
Done.

<details>
<summary>diff</summary>

```diff
diff --git a/include/small/ibuf.h b/include/small/ibuf.h
index 04c1568..1e9c848 100644
--- a/include/small/ibuf.h
+++ b/include/small/ibuf.h
@@ -33,14 +33,7 @@
 #include <stddef.h>
 #include <assert.h>
 
-#include "small_config.h"
-
-#ifdef ENABLE_ASAN
-#  include <sanitizer/asan_interface.h>
-#else
-#  define ASAN_POISON_MEMORY_REGION(addr, size) ((void)(addr), (void)(size))
-#  define ASAN_UNPOISON_MEMORY_REGION(addr, size) ((void)(addr), (void)(size))
-#endif
+#include "util.h"
 
 #if defined(__cplusplus)
 extern "C" {
diff --git a/include/small/util.h b/include/small/util.h
index 87a2489..845c7b1 100644
--- a/include/small/util.h
+++ b/include/small/util.h
@@ -250,4 +250,9 @@ small_asan_header_from_payload(void *payload)
 void
 small_asan_free(void *header);
 
-#endif /* ifdef ENABLE_ASAN */
+#else /* ifndef ENABLE_ASAN */
+
+#define ASAN_POISON_MEMORY_REGION(addr, size) ((void)(addr), (void)(size))
+#define ASAN_UNPOISON_MEMORY_REGION(addr, size) ((void)(addr), (void)(size))
+
+#endif /* ifndef ENABLE_ASAN */

```

</details>

Reply:
Done.

<details>
<summary>diff</summary>

```diff
diff --git a/include/small/mempool_asan.h b/include/small/mempool_asan.h
index 5b11353..e20808b 100644
--- a/include/small/mempool_asan.h
+++ b/include/small/mempool_asan.h
@@ -71,6 +71,14 @@ struct mempool {
 	size_t objcount;
 	/** Alignment of every allocation. */
 	size_t alignment;
+	/** Unique id among pools. */
+	unsigned int id;
+};
+
+/** Extra data associated with each mempool object allocation. */
+struct mempool_object {
+	/** id of the pool that this object belong to. */
+	unsigned int pool_id;
 };
 
 void
diff --git a/small/mempool_asan.c b/small/mempool_asan.c
index 96f0d91..6a5f8de 100644
--- a/small/mempool_asan.c
+++ b/small/mempool_asan.c
@@ -30,6 +30,9 @@
  */
 #include "mempool.h"
 
+/** id for the next created mempool. */
+static __thread uint32_t mempool_id_next;
+
 enum {
 	/** Maximum alignment for the ASAN mempool allocation. */
 	MEMPOOL_ASAN_MAX_ALIGNMENT = 4096,
@@ -48,6 +51,8 @@ mempool_create(struct mempool *pool, struct slab_cache *cache,
 	pool->alignment = 1 << (__builtin_ffs(objsize) - 1);
 	if (pool->alignment > MEMPOOL_ASAN_MAX_ALIGNMENT)
 		pool->alignment = MEMPOOL_ASAN_MAX_ALIGNMENT;
+
+	pool->id = mempool_id_next++;
 }
 
 void
@@ -57,12 +62,13 @@ mempool_destroy(struct mempool *pool)
 	memset(pool, 0, sizeof(*pool));
 }
 
-void *
+SMALL_NO_SANITIZE_ADDRESS void *
 mempool_alloc(struct mempool *pool)
 {
 
 	struct mempool_object *obj =
 			small_asan_alloc(pool->objsize, pool->alignment, 0);
+	obj->pool_id = pool->id;
 	pool->objcount++;
 
 	return small_asan_payload_from_header(obj);
@@ -72,6 +78,8 @@ SMALL_NO_SANITIZE_ADDRESS void
 mempool_free(struct mempool *pool, void *ptr)
 {
 	struct mempool_object *obj = small_asan_header_from_payload(ptr);
+	small_asan_assert(obj->pool_id == pool->id &&
+			  "object is from different pool");
 	pool->objcount--;
 
 	small_asan_free(obj);
diff --git a/test/mempool.c b/test/mempool.c
index 38fe3c6..07e5a11 100644
--- a/test/mempool.c
+++ b/test/mempool.c
@@ -118,9 +118,45 @@ mempool_align()
 	check_plan();
 }
 
+static char assert_msg_buf[128];
+
+static void
+on_assert_failure(const char *filename, int line, const char *funcname,
+		  const char *expr)
+{
+	(void)filename;
+	(void)line;
+	snprintf(assert_msg_buf, sizeof(assert_msg_buf), "%s in %s",
+		 expr, funcname);
+	small_on_assert_failure = small_on_assert_failure_default;
+}
+
+static void
+mempool_membership(void)
+{
+	plan(1);
+	header();
+
+	struct mempool pool1;
+	struct mempool pool2;
+
+	mempool_create(&pool1, &cache, 100);
+	mempool_create(&pool2, &cache, 200);
+	void *ptr = mempool_alloc(&pool1);
+	fail_unless(ptr != NULL);
+	small_on_assert_failure = on_assert_failure;
+	mempool_free(&pool2, ptr);
+	small_on_assert_failure = small_on_assert_failure_default;
+	ok(strstr(assert_msg_buf,
+		  "object is from different pool\" in mempool_free") != NULL);
+
+	footer();
+	check_plan();
+}
+
 int main()
 {
-	plan(2);
+	plan(3);
 	header();
 
 	seed = time(NULL);
@@ -145,6 +181,7 @@ int main()
 
 	mempool_basic();
 	mempool_align();
+	mempool_membership();
 
 	slab_cache_destroy(&cache);
 

```

</details>